File: stripe/src/main/java/com/stripe/android/GooglePayConfig.java
package com.stripe.android;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Locale;

import org.json.JSONException;
import org.json.JSONObject;

@SuppressWarnings("WeakerAccess")
public final class GooglePayConfig {
    @NonNull private final String mPublishableKey;
    @NonNull private final String mApiVersion;
    @Nullable private final String mConnectedAccountId;

    /**
     * Instantiate with {@link PaymentConfiguration}. {@link PaymentConfiguration} must be
     * initialized.
     */
    public GooglePayConfig(@NonNull Context context) {
        this(PaymentConfiguration.getInstance(context).getPublishableKey(), null);
    }

    /**
     * Instantiate with {@link PaymentConfiguration} and Connect Account Id.
     * {@link PaymentConfiguration} must be initialized.
     */
    public GooglePayConfig(@NonNull Context context, @Nullable String connectedAccountId) {
        this(PaymentConfiguration.getInstance(context).getPublishableKey(), connectedAccountId);
    }

    public GooglePayConfig(@NonNull String publishableKey) {
        this(publishableKey, null);
    }

    public GooglePayConfig(@NonNull String publishableKey, @Nullable String connectedAccountId) {
        mPublishableKey = ApiKeyValidator.get().requireValid(publishableKey);
        mApiVersion = ApiVersion.get().getCode();
        mConnectedAccountId = connectedAccountId;
    }

    /**
     * @return a {@link JSONObject} representing a
     * <a href="https://developers.google.com/pay/api/android/reference/object#gateway">
     *     Google Pay TokenizationSpecification</a> configured for Stripe
     */
    @NonNull
    public JSONObject getTokenizationSpecification() throws JSONException {
        return new JSONObject()
                .put("type", "PAYMENT_GATEWAY")
                .put(
                        "parameters",
                        new JSONObject()
                                .put("gateway", "stripe")
                                .put("stripe:version", mApiVersion)
                                .put("stripe:publishableKey", getKey())
                );
    }

    @NonNull
    private String getKey() {
        if (mConnectedAccountId != null) {
            return String.format(Locale.ROOT, "%s/%s", mPublishableKey, mConnectedAccountId);
        } else {
            return mPublishableKey;
        }
    }
}

File: stripe/src/main/java/com/stripe/android/GooglePayConfig.kt
package com.stripe.android

import android.content.Context
import org.json.JSONException
import org.json.JSONObject

/**
 * Configuration settings for Google Pay's `TokenizationSpecification`.
 */
class GooglePayConfig @JvmOverloads constructor(
    publishableKey: String,
    private val connectedAccountId: String? = null
) {
    private val validPublishableKey: String = ApiKeyValidator.get().requireValid(publishableKey)
    private val apiVersion: String = ApiVersion.get().code

    /**
     * @return a [JSONObject] representing a [Google Pay TokenizationSpecification](https://developers.google.com/pay/api/android/reference/object#gateway)
     * configured for Stripe
     */
    val tokenizationSpecification: JSONObject
        @Throws(JSONException::class)
        get() = JSONObject()
            .put("type", "PAYMENT_GATEWAY")
            .put(
                "parameters",
                JSONObject()
                    .put("gateway", "stripe")
                    .put("stripe:version", apiVersion)
                    .put("stripe:publishableKey", key)
            )

    private val key: String
        get() = connectedAccountId?.let { "$validPublishableKey/$it" } ?: validPublishableKey

    /**
     * Instantiate with [PaymentConfiguration].
     * [PaymentConfiguration] must be initialized.
     */
    constructor(context: Context) : this(
        PaymentConfiguration.getInstance(context)
    )

    /**
     * Instantiate with [PaymentConfiguration] and optional Connect Account Id.
     * [PaymentConfiguration] must be initialized.
     */
    @Deprecated("Configure connectedAccountId in PaymentConfiguration")
    constructor(context: Context, connectedAccountId: String? = null) : this(
        publishableKey = PaymentConfiguration.getInstance(context).publishableKey,
        connectedAccountId = connectedAccountId
    )

    private constructor(paymentConfiguration: PaymentConfiguration) : this(
        publishableKey = paymentConfiguration.publishableKey,
        connectedAccountId = paymentConfiguration.stripeAccountId
    )
}
